test/extensions/filters/http/ext_proc/ext_proc_grpc_fuzz.cc

```diff
@@ -0,0 +1,303 @@
+        uint32_t data_size = fdp->ConsumeIntegralInRange<uint32_t>(0, ext_proc_fuzz_max_data_size);
+        std::string data = std::string(data_size, 'a');
+        return sendDownstreamRequestWithBody(data, absl::nullopt);
+      } else {
+        ENVOY_LOG_MISC(trace, "Sending POST request with chunked body");
+        return sendDownstreamRequestWithChunks(fdp, fh, absl::nullopt);
+      }
+    default:
+      ENVOY_LOG_MISC(error, "Unhandled HttpMethod");
+      exit(EXIT_FAILURE);
+    }
+  }
+
+  envoy::extensions::filters::http::ext_proc::v3alpha::ExternalProcessor proto_config_{};
+  TestProcessor test_processor_;
+  Network::Address::IpVersion ip_version_;
+  Grpc::ClientType client_type_;
+};
+
+DEFINE_FUZZER(const uint8_t* buf, size_t len) {
+  // Split the buffer into two buffers with at least 1 byte
+  if (len < 2) {
+    return;
+  }
+
+  // External Process and downstream are on different threads so they should
+  // have separate data providers
+  size_t downstream_buf_len = len / 2;
+  size_t ext_proc_buf_len = len - downstream_buf_len;
+
+  // downstream buf starts at 0, ext_prob buf starts at buf[downstream_buf_len]
+  FuzzedDataProvider downstream_provider(buf, downstream_buf_len);
+  FuzzedDataProvider ext_proc_provider(&buf[downstream_buf_len], ext_proc_buf_len);
+
+  ExtProcIntegrationFuzz fuzzer(Network::Address::IpVersion::v4, Grpc::ClientType::GoogleGrpc);
+  ExtProcFuzzHelper fuzz_helper(&ext_proc_provider);
+
+  // This starts an external processor in a separate thread. This allows for the
+  // external process to consume messages in a loop without blocking the fuzz
+  // target from receiving the response.
+  fuzzer.test_processor_.start(
+      [&fuzz_helper](grpc::ServerReaderWriter<ProcessingResponse, ProcessingRequest>* stream) {
+        while (true) {
+          ProcessingRequest req;
+          if (!stream->Read(&req)) {
+            return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, "expected message");
+          }
+
+          // The following blocks generate random data for the 9 fields of the
+          // ProcessingResponse gRPC message
+
+          // 1 - 7. Randomize response
+          // If true, immediately close the connection with a random Grpc Status.
+          // Otherwise randomize the response
+          ProcessingResponse resp;
+          if (fuzz_helper.provider_->ConsumeBool()) {
+            return fuzz_helper.randomGrpcStatusWithMessage();
+          } else {
+            fuzz_helper.randomizeResponse(&resp, &req);
+          }
+
+          // 8. Randomize dynamic_metadata
+          /* TODO(ikepolinsky): Skipping - not implemented
+          if (fuzz_helper.provider_->ConsumeBool()) {
+          } */
+
+          // 9. Randomize mode_override
+          if (fuzz_helper.provider_->ConsumeBool()) {
+            ProcessingMode* msg = resp.mutable_mode_override();
+            fuzz_helper.randomizeOverrideResponse(msg);
```

Comment:
Because I'm using the FuzzedDataProvider to generate actions, adding more logging is definitely the way to go.

Reply:
OK. I think it's worth calling this out in your wrapup on this work, I see this as an important negative of FDP vs. protobuf, but not strong enough to avoid FDP.